### DISCUSSION TITLE
Section titles appear after tables (PDF)

### DISCUSSION THREAD

**Discussion by mattocci27** [quarto-dev/quarto-cli#2264] [Q&A]
Section titles appear after tables because Quarto produces tex files starting tables with `\begin{table}` instead of `\begin{table}[h!]`.  I can edit tex files but it is easier if Quarto can directly generate tex files starting tables with `\begin{table}[h!]`.  Are there any easy way to place section titles before tables?

Reference:
https://tex.stackexchange.com/questions/354897/section-title-appears-after-the-table

**Comment by cderv** [comment]
How are you creating your table in the Quarto document ? 

Just asking because it can be created different ways, including with Computation (like with a R package). Depending on the option, the tool that generate the LaTeX table is not the same so it would help know more about this. 

Thank you !

**Reply by mattocci27** [reply]
Yes, it might be an R specific problem.

This generates the section name in a weird place.

````
\newpage

# Table

```{r}
#| label: tbl-iris
#| tbl-cap: "Iris Data"

library(knitr)
kable(head(iris))
```
````



This works fine.

````
```
\newpage

# Table

| Col1 | Col2 | Col3 |
|------|------|------|
| A    | B    | C    |
| E    | F    | G    |
| A    | G    | G    |

: My Caption {#tbl-letters}

See @tbl-letters.
```
````

**Reply by cderv** [reply]
Thanks that helps a lot! 

When using `knitr::kable()` for PDF output, LaTeX table will be generated by **knitr** by default (not markdown). 

You can control the LaTeX behavior using argument in the function. `position = "h!` should allow you to add the `[h!]` after `\begin{table}`

More on configuration in https://bookdown.org/yihui/rmarkdown-cookbook/kable.html#customize-latex-tables looks for `position`

You could also force to use Markdown table using the `format` argument in `kable`. You may loose some nice **knitr** feature though.

**Reply by mattocci27** [reply]
Thanks! It worked without cross-reference but didn't work with cross-reference.

````
\newpage

# Tables

```{r}
#| label: tbl-iris
#| tbl-cap: "Iris Data"
library(knitr)
kable(head(iris), format = "latex", table.envir = "table", position = "h!")
```
````

 The above qmd code generates adds `[h!]` to a wrong place:

```
\begin{table}
\caption{\label{tbl-iris}Iris Data }\tabularnewline
[h!]
```

Is this a problem of `knitr::kable()`?

**Reply by cderv** [reply]
No this is something related to quarto.  We solved such type of issues in the past, but it seems we missed some edge case. 
Thanks for the report. I opened https://github.com/quarto-dev/quarto-cli/issues/2295 to track
